### ISSUE TITLE
Variable sinking rates in WOMBAT-lite

### ISSUE THREAD

**Issue by pearseb** [ACCESS-NRI/GFDL-generic-tracers#14]
Add a spatially variable sinking rate into WOMBAT-lite, although alternatives may need to be found, such as instead adding an explicit DOC (dissolved organic carbon) tracer.

**Comment by pearseb** [comment]
Starting this now. 

I was going to use BLING as the format to introduce the variable sinking scheme that has been used in the development of WOMBAT. However, it looks like BLING has variable sinking rates but does this implicitly rather then explicitly. They do not carry actively sinking tracers that are affected by the ocean circulation, which is what we want for WOMBAT.

This is a problem. There are two solutions as I see:

1. We may need to introduce a new tracer, called dissolved organic carbon (DOC), which accumulates in the gyres and in a sense is what the variable sinking scheme was originally designed to emulate. Low sinking rates in the gyres in the development of WOMBAT was intended to emulate the high amounts of suspended organic matter that exist in the gyres and is rapidly recycled by the plankton community here.
2. We make sinking speed constant, but we develop a complex representation of the remineralisation term that (i) increases with nutrient limitation (emulating bacteria outcompeting phytoplankton, or alternatively commensalism/mutualism between phytoplankton and bacteria) and (ii) decreases linearly with depth to emulate the increase in sinking speeds with depth.

**Comment by dougiesquire** [comment]
Another option could be to extend the generic tracer vertical diffusion routines to allow spatially varying sinking rates

**Comment by pearseb** [comment]
But vertical diffusion is not sinking right? Or is there where the sinking is handled by MOM?

**Comment by dougiesquire** [comment]
Yeah, the sinking is handled as part of the vertical tracer routines

**Comment by pearseb** [comment]
Hmmm... I will need help with this stuff. Can we talk to someone who is at work currently at ACCESS-NRI to help with this part of the dev? I can also reach out to people at CSIRO who are more numerically gifted than me or more familiar with the vertical tracer routines.

**Comment by pearseb** [comment]
I found something. This is in generic_tracers_utils.F90

I don't fully understand what's happening in this piece of code, but it's potentially where we would include variable sinking. 

![Image](https://github.com/user-attachments/assets/946f3bd7-e65b-47fc-978a-a9e6837e4134)

**Comment by dougiesquire** [comment]
Snap!

**Comment by pearseb** [comment]
Okay I'm going to move onto some revisions and writing, but this looks promising. I'd like to meet with you @dougiesquire and @chrisb13 at some point to see whether it would be possible to modify the _generic_tracer_utils.F90_ file so that we can implement variable sinking in MOM. 

Specific questions I have are:

1. How do we implement this so it is uploaded to the GFDL trunk with WOMBAT? We want WOMBAT to be included in GFDL's Git right?
2. How do we ensure that this works with MOM5 and MOM6? Particularly since the vertical spacing is so wacky in MOM6 with those µm thick layers.

**Comment by dougiesquire** [comment]
> I'd like to meet with you @dougiesquire and @chrisb13 at some point to see whether it would be possible to modify the generic_tracer_utils.F90 file so that we can implement variable sinking in MOM.

If we calculate the vertical movement from sinking within WOMBAT, I'm not sure we'd need to modify `generic_tracer_utils.F90` at all. But yes, happy to meet and chat.

**Comment by chrisb13** [comment]
Also happy to meet and chat.

**Comment by pearseb** [comment]
Let's go tomorrow? When's good for you Dougie?

**Comment by chrisb13** [comment]
I'm flexible but might have a meeting with the US (i.e. likely morning here). Afternoon is safe but I'm flexible for @dougiesquire.

**Comment by dougiesquire** [comment]
Apologies I can't do today and I'm away next week. Could do the week after

**Comment by pearseb** [comment]
Okay let's leave this for a moment and come back to it when Dougie's back. I have plenty to get on with in the meantime, although I am a bit nervous about the end of November coming up and having WOMBAT-lite working properly in generic tracers by the end of year.

**Comment by pearseb** [comment]
Update @dougiesquire @chrisb13.

I've managed to the get the "move_vertical" functionality working with WOMBATlite. My detritus tracer is now sinking using this function. 

However, the sinking tracer is not being deposited to the sediments at the bottom grid cell, and so it is accumulating to unrealistic values in the water column just above the sediments. Due to the complexity of the code, I might need some help figuring this out. I will continue to work on it in the meantime though.

Below is figure of what I'm talking about. You can see the accumulation of detritus in the bottom grid cell on shallow shelves. Note that the scale is a log10 scale, so the peak value there is 200 mmol/m3 of organic carbon, when typical values in productive waters are < 3.

![Image](https://github.com/user-attachments/assets/fa5d25c6-4583-4d25-82e0-6af30eb83e07)

**Comment by pearseb** [comment]
Looks like two different tri-diagonal solvers are called in MOM... 1 for when you do not have any vertical movement with this functionality, and another when you do. I'll have to alter them to make sure that stuff is being added to the sediments.

**Comment by pearseb** [comment]
I've implemented the change to the _g_tracer_vertdiff_M_ subroutine within the _generic_tracers_utils.F90_ file. It looks like this:

![Image](https://github.com/user-attachments/assets/71af1a14-45f3-45fb-aa01-57df1750c7d9)


**NOTE** that most of the changes above are just syntax repositioning of the lines. The actual changes in code are
![Image](https://github.com/user-attachments/assets/e47849c6-7faf-4bbd-8790-965910ccbc87)
![Image](https://github.com/user-attachments/assets/917448a2-8f81-4673-b08f-c3be46c56134)

and

![Image](https://github.com/user-attachments/assets/1d3146ed-e905-4dcc-b07f-6acb687128da)

These changes compiled successfully and I'm running tests.

**Comment by pearseb** [comment]
**Success!**

![Image](https://github.com/user-attachments/assets/f2b29a88-f8b6-47e6-a1b8-58d0620de375)

The bottom reservoir is accumulating and detritus is moving from the ocean into the sediments through the bottom of the ocean model.

**Comment by pearseb** [comment]
Takes ~40 minutes to run 1 year.


======================================================================================
                  Resource Usage on 2024-11-14 15:04:00:
   Job Id:             128602083.gadi-pbs
   Project:            p66
   Exit Status:        0
   Service Units:      371.04
   NCPUs Requested:    288                    NCPUs Used: 288
                                           CPU Time Used: 153:18:21
   Memory Requested:   1000.0GB              Memory Used: 202.77GB
   Walltime requested: 10:00:00            Walltime Used: 00:38:39
   JobFS requested:    600.0MB                JobFS used: 8.16MB
======================================================================================


Variable sinking rates working well. Passing detritus into bottom reservoir. All looking well. Will create a pull release.

**Comment by dougiesquire** [comment]
Closing as this was address in #25 and #15
